### ISSUE TITLE
Heterogenous maps fail silently in outputs

### ISSUE THREAD

**Issue by jbardin** [hashicorp/terraform#9568]
While terraform can technically accept maps with mixed types in a config, they cannot currently be interpreted by `github.com/hashicorp/hil`. 

In most cases this produces an error to explain the issue, for example:

```
* At column 3, line 1: map "var.somemap" does not have homogenous value types. found TypeList,
```

However, these fail silently when used in an output, and If we have a number of modules tied together via their outputs, this can be confusing to debug.

**Comment by tmclaugh** [comment]
I landed on #9458 while trying to debug an issue and I'm not sure if this bug explains my situation or not.  When trying to create a nested map the inner map (value of the outer map) is contained in a list.  This does not happen if I use the map() though.

main.tf
```
resource "null_resource" "n" {
  triggers {}
}

output "nested_map" {
  value = {
    foo = {
      a = "value1"
      b = "value2"
    }
    bar = {
      a = "value3"
      b = "value4"
    }
  }
}

output "nested_map_function" {
  value = "${
    map(
      "foo", map("a", "value1", "b", "value2"),
      "bar", map("a", "value3", "b", "value4")
    )
  }"
}
```

outputs:
```
nested_map = {
  bar = [map[b:value4 a:value3]]
  foo = [map[b:value2 a:value1]]
}
nested_map_function = {
  bar = map[a:value3 b:value4]
  foo = map[a:value1 b:value2]
}
```

outputs JSON:
```
{
    "nested_map": {
        "sensitive": false,
        "type": "map",
        "value": {
            "bar": [
                {
                    "a": "value3",
                    "b": "value4"
                }
            ],
            "foo": [
                {
                    "a": "value1",
                    "b": "value2"
                }
            ]
        }
    },
    "nested_map_function": {
        "sensitive": false,
        "type": "map",
        "value": {
            "bar": {
                "a": "value3",
                "b": "value4"
            },
            "foo": {
                "a": "value1",
                "b": "value2"
            }
        }
    }
}
```

**Comment by jbardin** [comment]
@tmclaugh,

Thanks for the info and the reproduction! This is a separate problem, so I opened #11115.

**Comment by tmclaugh** [comment]
Thanks! 👍

**Comment by apparentlymart** [comment]
I think this is actually a more general problem: errors when evaluating outputs are *never* surfaced in the UI, and instead just cause the output to be dropped silently.

**Comment by jbardin** [comment]
@apparentlymart: Yes. I think there's open issues about that in general.

I've found that having outputs fail silently was done on purpose, and we need to sort out when and where we can actually make them fail as expected.

**Comment by in4mer** [comment]
If you stick them into a null_data_source, they'll error

**Comment by Dominik-K** [comment]
Stumbled upon this silent fail in `output` some days ago. Is there any (discussion) progress on it?

I really would like to have failures in `output` variables be shown in the UI (and stop the processing). That would make debugging quite easier or even unnecessary as the modules will be correct in the first place (at development stage).

I already wanted to create a new issue about it, but then found this one. However, here's my minimal working example:

```hcl
variable "map" {
  type = "map"
  default = {
    key_a = "a"
    key_b = "b"
  }
}

# this fails as intended -> `aws_route53_record.test: key "c" does not exist in map var.map
resource "aws_route53_record" "test" {
  name = "${var.map["c"]}"
  type = "A"
  zone_id = "asdfasdf"
}
# comment until here and then ...

# ... this does NOT fail on `terraform plan`
# and is not output on `terraform apply`
# -> Is this intended?
output "this-is-not-output" {
  value = "${var.map["c"]}"
}

provider "aws" {
  region = "eu-west-1"
}
```

@jbardin Can you change the issue title to this more general topic or should this go to another (new?) issue?

**Comment by jbardin** [comment]
Hi @Dominik-K,

For reference, the root issue we're using for this is #9080.
We have the mechanisms in place now to handle output errors. It was a fairly invasive change, and is also a breaking change for existing configs, so enabling it has to wait for the 0.11 release.

**Comment by Dominik-K** [comment]
@jbardin Thanks for your fast answer.

**Comment by jbardin** [comment]
This is fixed by #16204, and now enabled in master
